### ISSUE TITLE
Issue with 8.0.0-RC

### ISSUE THREAD

**Issue by pambrose** [rjaros/kvision#537]
When trying to upgrade to 8.0.0-RC, I am seeing the same issue in the address-fullstack-ktor example and my own code. When I do a build on address-fullstack-ktor I get:

```
e: file:///git/kvision-examples/addressbook-fullstack-ktor/build/generated/ksp/metadata/commonMain/kotlin/com/example/AddressService.kt:10:8 Class 'AddressService' is not abstract and does not implement abstract member 'getAddressList'.
e: file:///git/kvision-examples/addressbook-fullstack-ktor/build/generated/ksp/metadata/commonMain/kotlin/com/example/ProfileService.kt:10:8 Class 'ProfileService' is not abstract and does not implement abstract member 'getProfile'.
e: file:///git/kvision-examples/addressbook-fullstack-ktor/build/generated/ksp/metadata/commonMain/kotlin/com/example/RegisterProfileService.kt:10:8 Class 'RegisterProfileService' is not abstract and does not implement abstract member 'registerProfile'
````

Thanks.

**Comment by rjaros** [comment]
That's strange. I've tested all the examples and they are building without problems on my machine. I've also upgraded some other apps without such issues as well.

**Comment by pambrose** [comment]
Interesting!! 
I tried Java 11, 17, and 21 and all 3 had the same issue. I invoked build from the CLI and IJ.
Let me try it on a different machine.

**Comment by rjaros** [comment]
Have you fresh cloned kvision-examples repo?

**Comment by pambrose** [comment]
I do. 
My original problem was on OSX. I just went to a Linux box and did a fresh clone and a build on address-fullstack-ktor and I got the same errors.
This is really bizarre!

**Comment by pambrose** [comment]
I got a similar error in numbers-fullstack-ktor:
e: file:///git/kvision-examples/numbers-fullstack-ktor/build/generated/ksp/metadata/commonMain/kotlin/com/example/NumberService.kt:10:8 Class 'NumberService' is not abstract and does not implement abstract member 'numberToWords'.

**Comment by rjaros** [comment]
Please compare your commands and console output:
```
rjaros@rjaros1:~/temp$ git clone git@github.com:rjaros/kvision-examples.git
Klonowanie do „kvision-examples”...
remote: Enumerating objects: 21159, done.
remote: Counting objects: 100% (2464/2464), done.
remote: Compressing objects: 100% (411/411), done.
remote: Total 21159 (delta 1996), reused 2440 (delta 1978), pack-reused 18695 (from 1)
Pobieranie obiektów: 100% (21159/21159), 185.51 MiB | 3.77 MiB/s, gotowe.
Rozwiązywanie delt: 100% (15137/15137), gotowe.
rjaros@rjaros1:~/temp$ cd kvision-examples/addressbook-fullstack-ktor
rjaros@rjaros1:~/temp/kvision-examples/addressbook-fullstack-ktor$ ./gradlew jar

> Configure project :
w: 'java' Gradle plugin is not compatible with 'org.jetbrains.kotlin.multiplatform' plugin.

Consider adding a new subproject with 'java' plugin where the KMP project is added as a dependency.


> Task :kotlinNpmInstall
(node:45014) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
warning gettext-extract > gettext-extractor > glob@7.2.3: Glob versions prior to v9 are no longer supported
warning gettext-extract > gettext-extractor > glob > inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
warning workspace-aggregator-27320d97-5362-4abe-80b9-51e5a17b1ff6 > addressbook-fullstack-ktor-test > karma > glob@7.2.3: Glob versions prior to v9 are no longer supported
warning workspace-aggregator-27320d97-5362-4abe-80b9-51e5a17b1ff6 > addressbook-fullstack-ktor-test > mocha > glob@8.1.0: Glob versions prior to v9 are no longer supported
warning workspace-aggregator-27320d97-5362-4abe-80b9-51e5a17b1ff6 > addressbook-fullstack-ktor-test > mocha > glob > inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
warning workspace-aggregator-27320d97-5362-4abe-80b9-51e5a17b1ff6 > addressbook-fullstack-ktor-test > karma-webpack > glob@7.2.3: Glob versions prior to v9 are no longer supported
warning workspace-aggregator-27320d97-5362-4abe-80b9-51e5a17b1ff6 > addressbook-fullstack-ktor-test > karma > rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
warning workspace-aggregator-27320d97-5362-4abe-80b9-51e5a17b1ff6 > addressbook-fullstack-ktor-test > karma > rimraf > glob@7.2.3: Glob versions prior to v9 are no longer supported
warning chart.js@4.4.3: The engine "pnpm" appears to be invalid.
warning " > style-loader@4.0.0" has unmet peer dependency "webpack@^5.27.0".
warning " > imports-loader@5.0.0" has unmet peer dependency "webpack@^5.0.0".
warning "gettext.js > po2json@1.0.0-beta-3" has unmet peer dependency "commander@^6.0.0".
warning "gettext.js > po2json@1.0.0-beta-3" has unmet peer dependency "gettext-parser@2.0.0".
warning "gettext.js > po2json@1.0.0-beta-3" has unmet peer dependency "gettext-to-messageformat@0.3.1".
warning " > karma-junit-reporter@2.0.1" has unmet peer dependency "karma@>=0.9".
warning Ignored scripts due to flag.

> Task :compileKotlinJvm
w: file:///opt/rjaros/temp/kvision-examples/addressbook-fullstack-ktor/src/jvmMain/kotlin/com/example/Service.kt:28:11 This code uses error suppression for 'ACTUAL_WITHOUT_EXPECT'. While it might compile and work, the compiler behavior is UNSPECIFIED and WON'T BE PRESERVED. Please report your use case to the Kotlin issue tracker instead: https://kotl.in/issue
w: file:///opt/rjaros/temp/kvision-examples/addressbook-fullstack-ktor/src/jvmMain/kotlin/com/example/Service.kt:148:11 This code uses error suppression for 'ACTUAL_WITHOUT_EXPECT'. While it might compile and work, the compiler behavior is UNSPECIFIED and WON'T BE PRESERVED. Please report your use case to the Kotlin issue tracker instead: https://kotl.in/issue
w: file:///opt/rjaros/temp/kvision-examples/addressbook-fullstack-ktor/src/jvmMain/kotlin/com/example/Service.kt:158:11 This code uses error suppression for 'ACTUAL_WITHOUT_EXPECT'. While it might compile and work, the compiler behavior is UNSPECIFIED and WON'T BE PRESERVED. Please report your use case to the Kotlin issue tracker instead: https://kotl.in/issue

> Task :jsBrowserProductionWebpack
asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  07ed6d5fa32c1d54af0f.ttf (414 KiB)

webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/


BUILD SUCCESSFUL in 48s
25 actionable tasks: 25 executed
rjaros@rjaros1:~/temp/kvision-examples/addressbook-fullstack-ktor$ java -jar build/libs/addressbook-fullstack-ktor-1.0.0-SNAPSHOT.jar 
[2024-08-22 19:12:24,103]-[main] INFO  Application - Autoreload is disabled because the development mode is off.
[2024-08-22 19:12:24,309]-[main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
[2024-08-22 19:12:24,413]-[main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Start completed.
[2024-08-22 19:12:24,737]-[main] INFO  Application - Application started in 0.656 seconds.
[2024-08-22 19:12:24,737]-[main] INFO  Application - Application started: io.ktor.server.application.Application@15eb0ae9
[2024-08-22 19:12:24,827]-[DefaultDispatcher-worker-1] INFO  Application - Responding at http://0.0.0.0:8080
^C[2024-08-22 19:12:28,264]-[KtorShutdownHook] INFO  Application - Application stopping: io.ktor.server.application.Application@15eb0ae9
[2024-08-22 19:12:28,272]-[KtorShutdownHook] INFO  Application - Application stopped: io.ktor.server.application.Application@15eb0ae9
rjaros@rjaros1:~/temp/kvision-examples/addressbook-fullstack-ktor$ 

```

**Comment by rjaros** [comment]
```
rjaros@rjaros1:~/temp/kvision-examples/addressbook-fullstack-ktor$ java -version
openjdk version "17.0.12" 2024-07-16
OpenJDK Runtime Environment (Red_Hat-17.0.12.0.7-2) (build 17.0.12+7)
OpenJDK 64-Bit Server VM (Red_Hat-17.0.12.0.7-2) (build 17.0.12+7, mixed mode, sharing)
```

**Comment by pambrose** [comment]
Sorry, I am a dope. I was using `build` and not jar. Sorry for wasting your time. All good.

**Comment by rjaros** [comment]
Still that's interesting why the `compileCommonMainKotlinMetadata` task (executed by `build`) is failing. It seem to fail on compiling Kotlin file, which obviously is correct.

**Comment by LiYing2010** [comment]
>  Still that's interesting why the compileCommonMainKotlinMetadata task (executed by build) is failing

looks like it's same with this issue
https://youtrack.jetbrains.com/issue/KT-59739

And I think we can change the generated code to fix this issue

AS-IS:
```
expect class AddressService : IAddressService
```

CHANGE-TO:
```
expect class AddressService : IAddressService {
    override suspend fun getAddressList(search: String?, types: String, sort: Sort): List<Address>
    override suspend fun addAddress(address: Address): Address
    override suspend fun updateAddress(address: Address): Address
    override suspend fun deleteAddress(id: Int): Boolean
}
```

see:
https://youtrack.jetbrains.com/issue/KT-59739#how-to-fix-your-code


if you agree with this solution, I think we should fix the code which generate these `expect class` source code

**Comment by rjaros** [comment]
Thanks. I already know the issue. It will be addressed in the next major release, because it will introduce breaking changes to existing apps (all services implementing remote interfaces will need to be fixed by adding `override` to all methods).

**Comment by LiYing2010** [comment]
>   all services implementing remote interfaces will need to be fixed by adding override to all methods

do you mean `actual class` or `expected class` ?

my proposal above means to change the `expected class`, generated by `kvision-gradle-plugin`,
in folder `{project-root}/build/generated/ksp/metadata/commonMain/kotlin/com/example/`

and I think this fix can be done in `kvision-gradle-plugin`


for "services implementing remote interfaces", I guess it's `actual class`,
but I think there is no changes need to be done for `actual class`
so, no user code will be affected

**Comment by LiYing2010** [comment]
BTW, I made a workaround for our project
and I think it's a very common case that user projects using `build` task in their CI env

my workaround will disable the failed task `compileCommonMainKotlinMetadata`, and make task `build` work fine again

`build.gradle.kts`:
```
tasks {
    // !!! this is a workaround for https://github.com/rjaros/kvision/issues/537 !!!
    // !!! TODO: remove these setting, after issues 537 get fixed !!!
    val disableInvalidTasks = register("disableInvalidTasks") {
        val invalidTasks = listOf(
            "compileCommonMainKotlinMetadata"
        )

        invalidTasks.forEach { taskName ->
            val task = findByName(taskName)
            if (task != null) {
                task.enabled = false
                println("INFO: task $taskName disabled")
            } else {
                println("WARNING: Unknown task $taskName")
            }
        }
    }

    build {
        dependsOn(disableInvalidTasks)
    }
}
```

**Comment by LiYing2010** [comment]
>  Unfortunately it will. But I made a mistake. 
> Not the `override` but the `actual` keyword will need to be added

I see

so the `actual` classes which implements `@KVService` interface in user code will need to add `actual` keyword to every function

looks like it's the only way to fix issue [KT-59739](https://youtrack.jetbrains.com/issue/KT-59739), because it's the final decision of language design
and they also mentioned this issue here:
https://kotlinlang.org/docs/k2-compiler-migration-guide.html#stricter-rules-for-abstract-members-in-expected-classes

but I think it's acceptable, because it's not a huge change, user can fix their code in a few minutes
 
how do you think about it?

**Comment by rjaros** [comment]
> but I think it's acceptable, because it's not a huge change, user can fix their code in a few minutes

It depends how many services and methods you have :-P

> how do you think about it?

As I said earlier, it will be fixed in KVision 9.0.0. That's how semantic versioning works when making incompatible changes.

**Comment by LiYing2010** [comment]
OK
thank you for explaining it
